### ISSUE TITLE
Shader compilation error with #version 300 es

### ISSUE THREAD

**Issue by diricxbart** [vispy/vispy#1850]
The below code generates a rectangular PNG image with red on the left side and blue on the right side. The shaders for this are  implemented using both the default shader version as well as 300 es.

```
from vispy import gloo
from vispy import app
from vispy import io
import numpy as np


vertex = """
precision mediump float;

attribute vec2 a_position;
varying vec2 normalizedCoord;

void main(void)
{
    gl_Position = vec4(a_position, 0.0, 1.0);
    normalizedCoord = (a_position * 0.5 + 0.5).xy;
}
"""   

fragment = """
precision mediump float;

varying vec2 normalizedCoord;

void main()
{
    if (normalizedCoord.x <= 0.5) {
        gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
    } else {
        gl_FragColor = vec4(0.0, 0.0, 1.0, 1.0);
    }
}

"""
    
vertex_300es = """
#version 300 es

precision mediump float;

in vec2 a_position;
out vec2 normalizedCoord;

void main(void)
{
    gl_Position = vec4(a_position, 0.0, 1.0);
    normalizedCoord = (a_position * 0.5 + 0.5).xy;
}
"""   

fragment_300es = """
#version 300 es

precision mediump float;

in vec2 normalizedCoord;
out vec4 fragmentColor;

void main()
{
    if (normalizedCoord.x <= 0.5) {
        fragmentColor = vec4(1.0, 0.0, 0.0, 1.0);
    } else {
        fragmentColor = vec4(0.0, 0.0, 1.0, 1.0);
    }
}

"""

class Canvas(app.Canvas):
    def __init__(self):
        h, w = (400, 800)
        
        app.Canvas.__init__(self, size=(w, h), keys='interactive')
        ps = self.pixel_scale

        self.program = gloo.Program(vert=vertex, frag=fragment)
        #self.program = gloo.Program(vert=vertex_300es, frag=fragment_300es)
        
        # gloo needs 32bit
        y = np.linspace(-1 + 2.0/h, 1, h)
        x = np.linspace(-1 + 2.0/w, 1, w)
        a, b = np.meshgrid(x, y)
        data = np.c_[a.flatten(), b.flatten()]
        self.program['a_position'] = data.astype('float32')

        self.show()

    def on_resize(self, event):
        width, height = event.size
        gloo.set_viewport(0, 0, width, height)

    def on_draw(self, event):
        gloo.clear('white')
        self.program.draw('points')

if __name__ == '__main__':
    img = Canvas().render()
    io.write_png("wonderful.png", img)
```
Now, if I change `self.program = gloo.Program(vert=vertex, frag=fragment)` into `self.program = gloo.Program(vert=vertex_300es, frag=fragment_300es)`, the program uses the 300 es shaders. However these don't seem to work, I'm getting the following error:
```
WARNING: Traceback (most recent call last):
  File "issue.py", line 98, in <module>
    img = Canvas().render()
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/app/canvas.py", line 538, in render
    self.events.draw()
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "issue.py", line 95, in on_draw
    self.program.draw('points')
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/program.py", line 533, in draw
    canvas.context.flush_commands()
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/glir.py", line 572, in flush
    self._shared.flush(parser)
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/glir.py", line 819, in parse
    self._parse(command)
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/glir.py", line 787, in _parse
    ob.set_data(*args)
  File "/root/.local/share/virtualenvs/git-xNkeUiKt/lib/python3.7/site-packages/vispy/gloo/glir.py", line 930, in set_data
    (self._target, errormsg))
RuntimeError: Shader compilation error in 35632:
    on line 5: error: No precision specified in this scope for type `vec2'
      in vec2 normalizedCoord;
    on line 6: error: No precision specified in this scope for type `vec4'
      out vec4 fragmentColor;
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (egl) at 0x7f54e784a0b8>> for DrawEvent
```
So it complains about the precision for vecX not being specified, but normally setting the precision for float also covers vecX. Setting it explicitly for vec2 / vec4 doesn't get rid of the error.

When I give an illegal shader version, I get as a message `Supported versions are: 1.10, 1.20, 1.30, 1.40, 1.00 ES, and 3.00 ES`, so the used 300 es version does seem to be supported by my environment...

**Comment by djhoese** [comment]
No idea off the top of my head, but what operating system and what GPU?

**Comment by diricxbart** [comment]
I'm using the environment described here: https://github.com/vispy/vispy/issues/1848

**Comment by djhoese** [comment]
This usually isn't something that VisPy actually controls. It is pretty much passed directly to your OpenGL drivers to compile.

**Comment by diricxbart** [comment]
I don't really think that's true, it seems that the precision modifiers are stripped [here](https://github.com/vispy/vispy/blob/master/vispy/gloo/glir.py#L622) and [here](https://github.com/vispy/vispy/blob/master/vispy/gloo/glir.py#L627)...

If I change the version from `#version 300 es` to `#version 140`, then the code executes correctly...
So it seems to me that the 300 es version is not properly parsed?

**Comment by djhoese** [comment]
Changing the version to `330` works for me. Unless my IDE's "Find" functionality is wrong, that `Supported versions` functionality error is not coming from VisPy. What I meant by VisPy controlling this was the actual compilation of the shader and the warnings/errors it produces. You are right that we strip the precision information off and I'm having trouble remembering why that is done.

**Comment by djhoese** [comment]
Looks like it was added from the beginning of the WebGL handling by @almarklein:

https://github.com/vispy/vispy/commit/f6b728047ba40a4b7861663fe729e24887e4392d

@almarklein Any idea why the precision information was stripped out of the shaders? Best I could think of after some research is that `highp` is an optional feature of the OpenGL ES 2.0 spec.

**Comment by almarklein** [comment]
I'm not sure to be honest. I suspect that some with some drivers at the time the glsl was not allowed to have precision qualifiers.

**Comment by djhoese** [comment]
@diricxbart As mentioned in my most recent comment on #1854, I don't think this is something VisPy can or should necessarily support. But I'd like to know more about what you are trying to do. Is there a reason you have to use the `es` version on GLSL? Do you *need* the `precision` specifiers? Are you targeting a mobile device and want to test your mobile compatible shaders on a desktop GPU?

As far as I can tell VisPy is supposed to be able to handle this by allowing you to do `from vispy.gloo.gl import use_gl; use_gl('es2')` but this apparently depends on the ES2 library (Angle/DirectX).

**Comment by bdiricx** [comment]
This is used in the context of pre-processing data for AI machine learning purposes, in the healthcare sector. Development for this is mainly done by using servers with significant GPU capabilities, such as NVidia V100 cards. The same algorithms also need to run on edge devices, which then typically run GLSL es. The goal of specifying the precision was to minimize the differences between the edge device results and the GPU server results...

But as I read in #1854, I'm not sure if the precision specifier is considered a no-op or not when the code would be run on a NVidia V100 with the EGL backend (like our experiment is now setup).

**Comment by djhoese** [comment]
I'm really not familiar with EGL. Is it strictly ES? If so, we can probably patch vispy to treat the EGL backend like an ES backend.

**Comment by diricxbart** [comment]
If you look [here](https://en.wikipedia.org/wiki/EGL_(API)), it states "EGL is an interface between Khronos rendering APIs (such as OpenGL, OpenGL ES or OpenVG) and the underlying native platform windowing system." So it's not strictly ES...

Couldn't you just detect the 'es' part in the version string and keep precision specifiers in this case?

**Comment by djhoese** [comment]
> Couldn't you just detect the 'es' part in the version string and keep precision specifiers in this case?

Probably, but that isn't how the code is structured right now. I was trying to see if it was possible to keep the design the original vispy creators had for this and understand why it is the way it is.

**Comment by larsoner** [comment]
Closing for #1854

**Comment by djhoese** [comment]
I've closed #1854 as I don't think it is the right solution for this. Seeing as we are approaching ~10 months since the last update on this I'm guess you either have your own work around or are not using VisPy for this or that it wasn't actually an issue. I was actually going to suggest just not including the precision specifier and comparing the differences. You mentioned using it for pre-processing for AI machine learning purposes which is not something VisPy is really meant to do and I would recommend using other GPU-centric **compute** technologies (OpenCL, CUDA, etc). VisPy is really only for visualization in my opinion.

If someone really wants to support this I think the right solution is to do better handling of the `use_gl` function (see above) and detecting what the backend + GL backend really want. Blindly adding or removing precision identifiers in the GLSL is likely only to create bugs and failed code. If not for the people running the ES code, then the people using VisPy as is. Overall I think depending on OpenGL ES for high precision computing is maybe not the best route and newer technologies should be leveraged.
